StyleCop.Analyzers/StyleCop.Analyzers/DocumentationRules/SA1620GenericTypeParameterDocumentationMustMatchTypeParameters.cs

```diff
@@ -51,7 +60,89 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
         /// <inheritdoc/>
         public override void Initialize(AnalysisContext context)
         {
-            // TODO: Implement analysis
+            context.RegisterSyntaxNodeActionHonorExclusions(this.HandleDocumentationTrivia, SyntaxKind.SingleLineDocumentationCommentTrivia);
+        }
+
+        private void HandleDocumentationTrivia(SyntaxNodeAnalysisContext context)
+        {
+            DocumentationCommentTriviaSyntax syntax = context.Node as DocumentationCommentTriviaSyntax;
+
+            // Find the type parameters of the parent node
+            IEnumerable<string> parentTypeParametersEnumerable = GetParentTypeParameters(syntax);
+
+            if (parentTypeParametersEnumerable == null)
+            {
+                return;
+            }
+
+            ImmutableArray<string> parentTypeParameters = parentTypeParametersEnumerable.ToImmutableArray();
+
+            ImmutableArray<XmlNodeSyntax> nodes = syntax.Content
+                .Where(node => string.Equals(GetName(node)?.ToString(), XmlCommentHelper.TypeParamXmlTag))
+                .ToImmutableArray();
+
+            for (int i = 0; i < nodes.Length; i++)
+            {
+                HandleElement(context, nodes[i], parentTypeParameters, i, GetName(nodes[i])?.GetLocation());
+            }
+        }
+
+        private static XmlNameSyntax GetName(XmlNodeSyntax element)
+        {
+            return (element as XmlElementSyntax)?.StartTag?.Name
+                ?? (element as XmlEmptyElementSyntax)?.Name;
+        }
+
+        private static void HandleElement(SyntaxNodeAnalysisContext context, XmlNodeSyntax element, ImmutableArray<string> parentTypeParameters, int index, Location alternativeDiagnosticLocation)
+        {
+            var nameAttribute = XmlCommentHelper.GetFirstAttributeOrDefault<XmlNameAttributeSyntax>(element);
+
+            // Make sure we ignore violations that should be reported by SA1613 instead.
+            if (string.IsNullOrWhiteSpace(nameAttribute?.Identifier?.Identifier.ValueText))
+            {
+                return;
+            }
+
+            if (!parentTypeParameters.Contains(nameAttribute.Identifier.Identifier.ValueText))
+            {
+                context.ReportDiagnostic(Diagnostic.Create(Descriptor, nameAttribute?.Identifier?.GetLocation() ?? alternativeDiagnosticLocation, string.Format(MissingTypeParamForDocumentation, nameAttribute.Identifier.Identifier.ValueText)));
+            }
+            else if (parentTypeParameters[index] != nameAttribute.Identifier.Identifier.ValueText)
```

Comment:
:exclamation: This goes :boom: if you have more `<typeparam>` elements than generic type parameters.